[Config] Generate JSON schema for config

GromNaN opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix https://github.com/symfony/symfony/issues/59603
| License       | MIT

Supersedes https://github.com/symfony/symfony/pull/59620, but generates nested objects instead of flattened definitions with a hash as id. This is easier to read.

Add a `JsonSchemaGenerator` class to the Config component that generates a JSON Schema from a config node tree, and a `JsonSchemaConfigDumpPass` compiler pass in FrameworkBundle that generates a combined `config/schema.json` file for all registered bundles during debug compilation (when `symfony/yaml` is available).

The generated JSON Schema can be used by IDEs for YAML config file autocompletion and validation.

valtzu committed:
Generate JSON schema for config

Make methods public

New implementation

More types

Fix type for canBeDisabled

Improve support for additionnal types

Cleanup

Use array instead of stdClass

Update schema version

Fix changelog

Fix anyOf

Fix default type for array

Fix CS

Add null to enum values

Add backed enum support to JsonSchemaGenerator

[FrameworkBundle] Add JsonSchemaConfigDumpPass to generate JSON Schema for YAML config

[Config] Compact generated JSON schema using \$defs to reduce repetition

Extract the scalar type pattern (["array","object","string","number","boolean","null"])
into a \$defs section and replace all occurrences with \$ref, reducing schema file size
significantly (−37% on the fixture, more on real projects with many scalar nodes).

[FrameworkBundle] Structure schema with \$defs and when@env properties

- Move each bundle/extension schema into \$defs[alias] and reference it with \$ref
  in root properties, so the scalar def from compact() is also applied inside
  bundle schemas (halving repetition)
- Add when@<env> properties for env-specific bundles; all-env bundles also appear
  inside when@<env> to allow per-environment overrides
- Update compact() to merge with pre-existing \$defs and skip own reserved def names
  when recursing, avoiding self-referencing \$refs

Fix when@env properties nesting: snapshot allProperties before appending when@ entries

when@env properties contain only env-specific bundles, not 'all' bundles

Revert "when@env properties contain only env-specific bundles, not 'all' bundles"

This reverts commit e4930e2bf3de6f37c77acf2a63489f05831af305.

GromNaN commented on src/Symfony/Component/Config/Tests/Definition/Dumper/JsonSchemaDumperTest.php:
```diff
@@ -0,0 +1,228 @@
+        ]);
+        $jsonSchema = json_encode($schema, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR)."\n";
+
+        if ($_ENV['TEST_GENERATE_FIXTURES'] ?? false) {
+            file_put_contents(__DIR__ . '/../Fixtures/Configuration/ExampleConfiguration.schema.json', $jsonSchema);
+            $this->markTestIncomplete('TEST_GENERATE_FIXTURES is set');
+        }
```
Same pattern proposed in https://github.com/symfony/symfony/pull/62121

GromNaN commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+
+            if ($node->isDeprecated()) {
+                $schema['deprecated'] = true;
+                $schema['deprecationMessage'] = $node->getDeprecationMessage($node);
```
`deprecationMessage` is not a JSON Schema key. This info should move to `description`.

GromNaN replied:
Resolved by updating the json schema spec version: https://github.com/symfony/symfony/pull/62125#discussion_r2665464585

stof commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+    {
+        return [
+            '$schema' => 'http://json-schema.org/draft-07/schema',
+            ... $meta,
```
```suggestion
            ...$meta,
```

We don't put spaces after unary operators.

stof commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+            }
+
+            if (isset($schema['properties'])) {
+                if ($node->shouldIgnoreExtraKeys()) {
```
this could be simplified as `$schema['additionalProperties'] = $node->shouldIgnoreExtraKeys()` instead

stof commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+        } elseif ($node instanceof BooleanNode) {
+            $schema = ['type' => ['boolean']];
+
+            if ($node->isNullable()) {
```
don't we need the same for other kinds of nullable nodes ? And for array nodes that have a `treatNullLike*` config ?

mdeboer replied:
@stof the thing is, I saw a similar implementation in the current PHP reference file generator (or array shape generator?) where this was only handled for booleans, but not for strings for example. But as this was done so specifically for booleans, I thought it must have been intentional.

Judging by the implementation of the nodes itself, it seems like you mention, many nodes allow for nullable values in some shape or form.

It would probably make sense to have a similar implementation in both places when it comes to this.

GromNaN replied:
The `BooleanNode` is the only to have a `nullable` property. This needs more investigation.

mdeboer replied:
Given the latest changes, is this (still) blocking?

GromNaN replied:
This is addressed in the refactored `JsonSchemaDumper`: `BaseNode::isNullable()` was added and all node types now reflect nullable types in the generated schema (commit d9ea70beb78). `treatNullLike()` on array nodes is also covered since it maps `null` in `equivalentValues`, which `isNullable()` checks.

stof commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+                '$anyOf' => [
+                    ['enum' => array_map(static function (mixed $case): string|int|float|bool|null {
+                        if ($case instanceof \UnitEnum) {
+                            return \sprintf('!php/enum %s::%s', $case::class, $case->name);
```
this looks weird to me as it would define it as supported as a string value (the combination of custom Yaml tags with a json schema is an unknown case for me though)

And this misses the case of enum nodes with an enum class being a BackedEnum, where backing values are accepted in the input (and automatically converted to the corresponding case)

GromNaN replied:
JSON Schema doesn't support YAML tags. This is a slightly unusual use because it allows the string to be validated for the tag without the quotes. Of course, the version with the quotes is invalid, but the schema has no way of knowing this.

stof commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+                                        ExprBuilder::TYPE_INT => 'integer',
+                                        ExprBuilder::TYPE_STRING => 'string',
+                                        ExprBuilder::TYPE_BOOL => 'boolean',
+                                        ExprBuilder::TYPE_ARRAY => 'object',
```
this looks wrong to me. A PHP array can be either a JSON object or a JSON array.

GromNaN replied:
I added `['array', 'object']`, but I need to find the test case.

mdeboer commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+    public function generate(NodeInterface $node, array $meta = []): array
+    {
+        return [
+            '$schema' => 'http://json-schema.org/draft-07/schema',
```
With [opis/json-schema](https://github.com/opis/json-schema) and many other tools supporting [draft schema 2020-12](https://json-schema.org/draft/2020-12/schema) and considering that schema version has been around for a while, why was chosen not output the newest version?

I see for example the `deprecated` property is being used, which is only introduced starting from `2019-09`, not in `draft-07` (see [here](https://www.learnjsonschema.com/2020-12/meta-data/deprecated/)). Therefore the generated schema at this point wouldn't strictly be valid in `draft-07`.

stof commented on src/Symfony/Component/Config/CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ CHANGELOG
  * Deprecate accessing the internal scope of the loader in PHP config files, use only its public API instead
  * Deprecate setting a default value to a node that is required, and vice versa
  * Deprecate fluent config builders, return PHP arrays from your config instead
+ * Add `JsonSchemaGenerator` to generate JSON Schema from configuration definitions
```
should be in the `8.1` section.

mdeboer replied:
And the PR description should probably be updated too, to reference 8.1 instead of 7.4 as target.

mdeboer commented on src/Symfony/Component/Config/Builder/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,186 @@
+            $schema = ['type' => ['string']];
+        } elseif ($node instanceof EnumNode) {
+            $schema = [
+                '$anyOf' => [
```
This doesn't look right to me, not only should it be [`oneOf`](https://www.learnjsonschema.com/draft7/validation/oneof/) (not `$anyOf`, `anyOf` or `$anyOf`) as something can't be `null` _and_ a value of an enum.

But also because the [documentation](https://www.learnjsonschema.com/draft7/validation/enum/) mentions that `enum` can contain literal `null` to allow a `null` value, as is demonstrated in [this example](https://json-schema.org/understanding-json-schema/reference/enum#extended-example:-accepting-multiple-data-types)

GromNaN replied:
Thanks, that's more simple with `null` in the list of allowed values.

GromNaN commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 8.1
 ---
 
+ * Generate JSON schema for YAML configuration of bundles
```
This part was removed, it needs to be implemented in a compiler pass, in the same way as `PhpConfigReferenceDumpPass`.

GromNaN committed:
Add patternProperties for when@<env> and restrict known envs with additionalProperties

GromNaN committed:
Use precise regex ^when@[a-zA-Z0-9]+$ for patternProperties

GromNaN committed:
Remove compact() from JsonSchemaGenerator; revert services.schema.json merge

GromNaN committed:
Add BaseNode::isNullable() and reflect nullable types in JSON schema

Add a public isNullable() method to BaseNode that checks equivalentValues
for null-from entries. Use it in JsonSchemaGenerator to include "null" in
the type of array nodes that accept null as input (all builder-created
array/prototyped-array nodes, since ArrayNodeDefinition registers null => []
by default).

GromNaN committed:
Move JsonSchemaGenerator to Definition/Dumper/JsonSchemaDumper

GromNaN committed:
Use shared \$defs to reduce repetition in generated JSON schemas

- Add STRUCTURAL_DEFS (array, object, scalar, variable and their null
  variants) and PARAMETERIZABLE_DEFS (string, boolean, integer, number
  and their null variants) as private constants in JsonSchemaDumper
- dumpNode() now emits \$ref to \$defs/types/* instead of inline type
  schemas, keeping generated schemas concise
- ScalarNode (string|number|boolean|null) is now handled separately from
  VariableNode (any type), fixing overly-permissive schemas for scalar
  prototypes (e.g. MonologBundle channels)
- parameterSchemas are injected via constructor and baked directly into
  the typed defs (boolean, integer, number, string) via buildTypeDefs(),
  keeping the DI-specific %param% concept out of the symfony/config core
- getAllDefs() exposes {types: {...}} for callers building full schemas
- JsonSchemaConfigDumpPass nests type defs under \$defs/types/ (including
  the DI-specific param def) and bundle schemas under \$defs/nodes/

GromNaN committed:
Update CHANGELOG for JsonSchemaDumper

GromNaN committed:
Add flat anyOf fix and regression test for parameterSchemas + normalizedTypes

When a node schema already contains anyOf (e.g. EnumNode with parameterSchemas),
applying normalizedTypes must append to the existing anyOf rather than wrapping
it, to avoid nested anyOf structures.

GromNaN commented on src/Symfony/Component/Config/Definition/Dumper/JsonSchemaDumper.php:
```diff
@@ -0,0 +1,242 @@
+ */
+final class JsonSchemaDumper
+{
+    private const PARAMETERIZABLE_DEFS = [
+        'string' => ['type' => 'string'],
+        'string_null' => ['type' => ['string', 'null']],
+        'boolean' => ['type' => 'boolean'],
+        'boolean_null' => ['type' => ['boolean', 'null']],
+        'integer' => ['type' => 'integer'],
+        'integer_null' => ['type' => ['integer', 'null']],
+        'number' => ['type' => 'number'],
+        'number_null' => ['type' => ['number', 'null']],
+    ];
+
+    private const STRUCTURAL_DEFS = [
+        'variable' => ['type' => ['array', 'boolean', 'null', 'number', 'object', 'string']],
+        'scalar' => ['type' => ['boolean', 'null', 'number', 'string']],
+        'array' => ['type' => 'array'],
+        'array_null' => ['type' => ['array', 'null']],
+        'object' => ['type' => 'object'],
+        'object_null' => ['type' => ['object', 'null']],
+    ];
```
Using this predefined type definitions is an optimization of the size of the schema file. For Symfony/Demo, the `schema.json` file goes from 12k lines to 8k thanks to this.

nicolas-grekas commented:
I went through the five open threads against the current head, running the code rather than reading it. Three are resolved by the April rework and can be closed:

* @mdeboer's draft version point: the dumper now declares `https://json-schema.org/draft/2020-12/schema`, where `deprecated` is valid;
* @stof's `BackedEnum` point: backing values are emitted next to the `!php/enum` form;
* the boolean-only nullability special case: `BaseNode::isNullable()` now generalises it, and the object, array and boolean branches consult it.

But the nullability discussion @stof started is only half done. I built a tree with one node of each kind and compared what the node accepts at runtime (`normalize()` then `finalize()`) with what the generated schema says:

```
node                                 runtime(null)   schema
stringNode('s')                      accepted        {"$ref": ".../string"}          too strict
stringNode('s')->treatNullLike('')   accepted        {"$ref": ".../string"}          too strict
integerNode('i')                     REJECTED        {"$ref": ".../integer"}         consistent
integerNode('i')->treatNullLike(0)   accepted        {"$ref": ".../integer"}         too strict
enumNode('e')->values(['a','b'])     REJECTED        {"enum": ["a","b",null]}        too lax
  ... ->values(['a','b',null])       accepted        {"enum": ["a","b",null,null]}   double null
  ... ->enumFqcn(Suit::class)        accepted        {"enum": [...cases..., null]}   consistent
```

Three findings:

* the `string_null`, `integer_null` and `number_null` `$defs` are defined but never referenced: the `StringNode` and `NumericNode` branches do not consult `isNullable()`, unlike their boolean, object and array siblings. With `treatNullLike()`, the IDE would flag `~` as invalid on a node that accepts it;
* `EnumNode` appends `null` to every enum unconditionally. A plain value enum rejects null at runtime (`The value of type "null" is not allowed for path "mode"`), so the schema validates configs the component throws on. When null *is* a declared value it is emitted twice. Only the `enumFqcn` case accepts null implicitly, so the append should be gated on `\in_array(null, $values, true)` or the enum-FQCN case rather than unconditional;
* a plain `StringNode` accepts null at runtime through its `ScalarNode` heritage, while `isNullable()` reports false, since it only looks at `equivalentValues`. So even wiring `isNullable()` into the string branch would not close the gap. That is worth a decision rather than a patch: either `isNullable()` should reflect `validateType()`'s actual null acceptance, or the schema knowingly stays stricter than the runtime for plain scalars, which is arguably the better DX and just deserves a comment saying so.

One note on my harness in case you want to reproduce: `finalize(null)` alone reports `treatNullLike()` nodes as rejecting, because that conversion happens in `normalize()`. The table above runs both phases.

nicolas-grekas commented:
I pushed the fixes for my findings above as a commit on top, so both existing authorships stay intact, and resolved the five open threads: three were already answered by the April rework (draft 2020-12, `BackedEnum` backing values, `isNullable()` replacing the boolean special case), and the remaining nullability gaps are closed by the new commit:

* the `StringNode` and `NumericNode` branches now consult `isNullable()`, so the `string_null`, `integer_null` and `number_null` `$defs` are referenced at last;
* `EnumNode` no longer appends `null` unconditionally: it is emitted when null is a declared value or when the node is bound to an enum FQCN, which is the case that accepts and returns null at runtime;
* `BaseNode::isNullable()` no longer counts the identity null mapping the builder registers on every node. That mapping leaves null to be rejected by the type check, so an `integerNode()` without `treatNullLike()` reported nullable while rejecting null.

Two behaviour notes visible in the regenerated `ExampleConfiguration.schema.json`:

* a plain `stringNode()` is now `string_null`, because `StringNodeDefinition` defaults `nullEquivalent` to `''`: writing `~` on such a node is accepted and means the empty string;
* `string_list_default_null`, an `arrayNode()->defaultNull()`, went from `array_null` to `array`: `defaultNull()` resets the null equivalence, and writing `~` explicitly on that node throws `Expected "array", but got "null"`. The schema now matches that.

Every node kind and null-handling combination is asserted against what `normalize()` plus `finalize()` actually accepts; reverting the two source files fails 8 of the test rows. Config is green at 614 tests and the FrameworkBundle compiler tests at 18.

GromNaN committed:
[Config] Generate a JSON schema from a configuration tree

JsonSchemaDumper turns a node tree into a JSON Schema (draft 2020-12), and
FrameworkBundle dumps a combined config/schema.json for all registered bundles
during debug compilation, so IDEs can validate and complete YAML configuration.
The schema states exactly what each node accepts: nullability follows the null
equivalences a node actually applies, and enums list their values, backing
values and null only when the node takes them.

Based on the initial implementation by valtzu in #59620.

nicolas-grekas commented:
The three red jobs after the squash were all this PR's, now fixed in the same commit:

* **low-deps and x86**: `JsonSchemaConfigDumpPass` lives in FrameworkBundle and instantiates `JsonSchemaDumper`, which is new in Config, so those jobs resolve `symfony/config` from packagist and every debug kernel boot died with `Class "JsonSchemaDumper" not found`. The pass now returns early when the class is missing, the same way it already does for `Yaml`, and its test skips in that situation instead of failing;
* **PHPStan**: `Cannot unset offset 'type'` on the schema shape. The `unset` was dead code, since `type` is only ever set non-empty (the fallback branch and the normalized-types block both write non-empty values), so the single-element collapse is now `\is_array($schema['type'] ?? null) && 1 === \count(...)` and the unset is gone.

Config is green at 622 tests, the compiler tests at 18, php-cs-fixer clean.

nicolas-grekas commented:
Thank you @GromNaN.

mdeboer commented:
Nice, really happy this landed in 8.2! 🚀 Thanks @GromNaN, @nicolas-grekas and @stof!

GromNaN commented:
I asked JetBrains to detect the schema file and apply it automatically: https://intellij-support.jetbrains.com/hc/en-us/community/posts/38145537718674-Auto-detect-Symfony-configuration-JSON-Schemas-in-PhpStorm

andrew-demb commented:
@mdeboer https://youtrack.jetbrains.com/issue/WI-85802/Auto-detect-Symfony-configuration-JSON-Schemas-in-PhpStorm#focus=Comments-27-14326617.0-0

https://github.com/SchemaStore/schemastore/pull/6225

mdeboer commented:
@GromNaN @andrew-demb woops my bad, great work! ❤️
